### ISSUE TITLE
When Handler Isn't Enough - Possible Improvements

### ISSUE THREAD

**Issue by mruoss** [mtrudel/thousand_island#144]
Hey Mat

If I read the docs correctly, `handle_connection/2` is called for every client connection. I'm looking for a callback to do some work in my handler when the process is initialized. However, GenServer's `init/1` callback is already implemented in the `__using__` macro and it's not overridable. Do you think you could offer a callback that is called from within `init/1` or so?

Best,
Michael

**Comment by mruoss** [comment]
Oh wait... I got it now. Every client spins up a new handler process of course. So `handle_connection/2` is equivalent to `init/1` in that sense, right?

**Comment by mtrudel** [comment]
Yep, that's exactly it. There's a fair bit of setup (mostly handshaking, control process management and telemetry setup) that needs to be done before a socket is 'ready for use' in the way that 99% of users think about being ready. If you *really* need to remove the guardrails, there's always the 'When Handler Isn't Enough' in https://hexdocs.pm/thousand_island/ThousandIsland.Handler.html, though you'll be on your own for all of the setup details. 

To date, I'm not actually aware of anyone ever having used that escape hatch in any substantial capacity; unless you're absolutely positively sure that you need it, `handle_connection/2` is probably the callback for you.

**Comment by mruoss** [comment]
Yeah I see. I think in my case it is actually _too much_ 😆. I just want to proxy the binaries so probably Thousand Island is a bit of an overkill and I can just open a port and take care of message passing...

**Comment by mruoss** [comment]
Okay... 2 days later and I am still here 😂 

Got a my POC working but I'd like to get rid of one of the processes because it's not actually needed. So... I'm pretty much positively sure that I want that escape hatch. Looking at `ThousandIsland.Handler.__using__` however scares me a bit. There's a lot going on in there and I'd have to copy paste most of it becuase I can't even `defdelegate` or forward to those functions. 

So here's my question: Why don't you put the implementations of the `def` inside that `__using__` macro out to `ThousandIsland.Handler` so it can be reused? This is actually a pretty common approach and would also make the modules that `use ThousandIsland.Handler` smaller.

**Quick example to show what I mean:**

```ex
defmodule ThousandIsland.Handler do
  defmacro __using__(opts) do
    quote location: :keep do
      def handle_info(
            {:thousand_island_ready, raw_socket, server_config, acceptor_span, start_time},
            {nil, state}
          ) do
        ThousandIsland.Handler.thousand_island_ready(__MODULE__, raw_socket, server_config, acceptor_span, start_time, state)
      end
    end
  end

  def thousand_island_ready(handler, raw_socket, server_config, acceptor_span, start_time, state) do
    # actual implementation
  end
end
```

**Comment by mtrudel** [comment]
Resolved in #146
